docs: Prepare Kafka v.4 initial documentation (#403)
canonical/kafka-operator: docs/tutorial/deploy.md

This is a part of the [Charmed Apache Kafka Tutorial](index.md).

## Deploy Charmed Apache Kafka

To deploy Charmed Apache Kafka, all you need to do is run the following commands, which will automatically fetch [Apache Kafka](https://charmhub.io/kafka?channel=4/edge) from [Charmhub](https://charmhub.io/) and deploy it to your model.

For example, to deploy a cluster of three Apache Kafka brokers, you can simply run:

```shell

> $ juju deploy kafka -n 3 --channel 4/edge --roles=broker

Reviewer: Chore: We should avoid using the `$` prompt (or any other prompt) in the console/shell/bash code blocks.

```{note}
KRaft replaces the dependency on Apache ZooKeeper for metadata management. For more information on the differences between the two solutions, please refer to the [upstream Apache Kafka documentation](https://kafka.apache.org/40/documentation/zk2kraft.html)
```

Charmed Apache Kafka can run both with `roles=broker` and/or `roles=controller`. To deploy a cluster of three KRaft controllers, run:

```shell
$ juju deploy kafka -n 3 --channel 4/edge --roles=controller controller
```

After this, it is necessary to connect the two clusters, taking care to specify which cluster is the orchestrator:

```shell
juju integrate kafka:peer-cluster-orchestrator kraft:peer-cluster
```

Juju will now fetch Charmed Apache Kafka and begin deploying both applications to the LXD cloud before connecting them to exchange access credentials and machine endpoints. This process can take several minutes depending on the resources available on your machine. You can track the progress by running:

```shell

> juju status --watch 1s

Reviewer: chore: We need to update to the watch command, instead of `--watch` parameter

Author: It's not deprecated yet is it?

Reviewer: It is not deprecated, at least yet, but someone said there is a memory leak when we use it with `--watch`. So the recommended approach is to use the watch command instead.

Author: Actually I'll update them all.

kraft  4.0.0    active      3  kafka  4/edge   226  no       

Unit      Workload  Agent  Machine  Public address  Ports      Message
kafka/0*  active    idle   0        10.233.204.241  19093/tcp  
kafka/1   active    idle   1        10.233.204.196  19093/tcp  
kafka/2   active    idle   2        10.233.204.148  19093/tcp  
kraft/0   active    idle   3        10.233.204.125  9098/tcp   
kraft/1*  active    idle   4        10.233.204.36   9098/tcp   
kraft/2   active    idle   5        10.233.204.225  9098/tcp   

Machine  State    Address         Inst id        Base          AZ  Message
0        started  10.233.204.241  juju-07a730-0  ubuntu@24.04      Running
1        started  10.233.204.196  juju-07a730-1  ubuntu@24.04      Running
2        started  10.233.204.148  juju-07a730-2  ubuntu@24.04      Running
3        started  10.233.204.125  juju-07a730-3  ubuntu@24.04      Running
4        started  10.233.204.36   juju-07a730-4  ubuntu@24.04      Running
5        started  10.233.204.225  juju-07a730-5  ubuntu@24.04      Running
```

To exit the screen with `juju status --watch 1s`, enter `Ctrl+c`.

## Access Apache Kafka brokers

> To watch the process, `juju status` can be used. Once all the units show as `active|idle`, the credentials are available to be retrieved.

Reviewer: Suggestion: Remove redundancy and improve wording.

```suggestion
Once all the units are shown as `active|idle`, the credentials can be retrieved.
```

> All sensitive configuration data that Charmed Apache Kafka uses such as passwords and SSL certificates is stored in Juju secrets, you can learn more about [Juju secrets in the documentation](https://documentation.ubuntu.com/juju/3.6/reference/secret/).

Reviewer: Suggestion: Minor wording improvements.

```suggestion
All sensitive configuration data used by Charmed Apache Kafka, such as passwords and SSL certificates, is stored in Juju secrets. See the [Juju secrets documentation](https://documentation.ubuntu.com/juju/3.6/reference/secret/) for more information.
```

```shell
d2lj5jgco3bs3dacm2tg:
  revision: 1
  checksum: a6517abdd5e22038bfafe988e6253bb03c0462067b50475789eb6bc658ee0b11
  owner: kafka
  label: cluster.kafka.app
  created: 2025-08-24T15:42:13Z
  updated: 2025-08-24T15:42:13Z
  content:
    admin-password: dxpex3Uc1sWIBna83gELtJOhAuW2awji
    sync-password: eqI0RLV1lRSaIIiDKf3yz0W66ajICmDT
    internal-ca: |-
      <multi-line-certificate>
    internal-ca-key: |-
      <multi-line-private-key>
```

The important line here for accessing the Apache Kafka cluster itself is `admin-password`, which tells us that `username=admin` and `password=dxpex3Uc1sWIBna83gELtJOhAuW2awji`. These are the credentials to use to successfully authenticate to the cluster.

For simplicity, the password can also be directly retrieved by parsing the YAML response from the previous command directly using `yq`:

```shell
juju show-secret --reveal cluster.kafka.app | yq '.. | ."admin-password"? // empty' | tr -d '"'
```

```{caution}
When no other application is integrated to Charmed Apache Kafka, the cluster is secured-by-default and external listeners (bound to port `9092`) are disabled, thus preventing any external incoming connection. 
```

> We will also need a 'bootstrap server' Apache Kafka broker address and port to initially connect to. When any application first connects to a `bootstrap-server`, the client will automatically make a metadata request that returns the full set of Apache Kafka brokers with their addresses and ports.

Reviewer: Suggestion: Minor wording improvements.

```suggestion
We will also need a bootstrap server Apache Kafka broker address and port to initially connect to. When any application connects for the first time to a `bootstrap-server`, the client will automatically make a metadata request that returns the full set of Apache Kafka brokers with their addresses and ports.
```

> To use `kafka/0` as the `bootstrap-server`, retrieve it's IP address and add a port with:

Reviewer: Todo: typo fix.

```suggestion
To use `kafka/0` as the `bootstrap-server`, retrieve its IP address and add a port with:
```
